### DISCUSSION TITLE
YAML Schemas

### DISCUSSION THREAD

**Discussion by ssbanerje** [LunarVim/LunarVim#1753] [Q&A] [answered]
How does one set the YAML schema for a buffer?

It seems that the `schemaStore` that is used in the default LVIM config loads several schemas. But the file matching happens based on the order in which the YAML schemas are loaded.

For example, when opening `.github/workflows/build.yaml`, the schema picked is `hammerkit` instead of GitHub Workflows/actions.

Seems like this entry in schema store
```json
{
      "name": "Hammerkit YAML Schema",
      "description": "JSON Schema for hammerkit files.",
      "fileMatch": [
        "build.yaml",
        "build.yml"
      ],
      "url": "https://json.schemastore.org/hammerkit.json"
}
```
comes before the GitHub entry
```json
{
      "name": "GitHub Workflow",
      "description": "YAML schema for GitHub Workflow",
      "fileMatch": [
        ".github/workflows/**.yml",
        ".github/workflows/**.yaml"
      ],
      "url": "https://json.schemastore.org/github-workflow.json"
}
```

How does one locally set this?

The same thing happens with the Kubernetes settings here:
https://github.com/LunarVim/LunarVim/blob/9e18c16b2b378792c689591190774786fd5da04c/lua/lvim/lsp/providers/yamlls.lua#L11-L25

All of these files (even the wildcard matches for example `.github/workflows/deployment.yaml`) are always recognized as K8s schema.

**Comment by abzcoding** [comment]
yeah this is my fault, because most of my workflow was around k8s, i think we need to remove these from the default lvim config and let users add their own schemas based on their workflow

**Reply by abzcoding** [reply]
cc: @ChristianChiarulli @kylo252

**Reply by ssbanerje** [reply]
Is there a way for users to set this per project or per buffer?

I think the problem is that the greedy regex matching for the filenames in the schemastore will always lead to this issue.

In the long term there must be a way for a user to specify a priority order or something on which the schemas are evaluated. That way the most relevant schemas are evaluated first.

**Reply by abzcoding** [reply]
I think the best way to is to remove hardcoded default configurations,
then let users override and write their own provider
for example https://github.com/abzcoding/lvim/tree/main/lua/user/providers
and just document this process in lunarvim.org
and about the project-based mode, you might be able to that with some logic in your provider config
for example, see this provider https://github.com/abzcoding/lvim/blob/main/lua/user/providers/texlab.lua
which uses this configuration https://github.com/abzcoding/lvim/blob/main/lua/user/tex.lua

**Reply by ssbanerje** [reply]
Ok... in the meantime I think I have found a workaround. The yamlls allows specifying a modeline:
```yaml
# yaml-language-server: $schema=<urlToTheSchema>
```
Pointing this to the right schemastore.org URL works.

**Reply by abzcoding** [reply]
you mean this url is wrong?

**Reply by abzcoding** [reply]
https://github.com/LunarVim/LunarVim/blob/9e18c16b2b378792c689591190774786fd5da04c/lua/lvim/lsp/providers/yamlls.lua#L8-L10

**Reply by ssbanerje** [reply]
No... putting the comment at the top of a file with a URL like https://json.schemastore.org/github-workflow.json, will make the LS use that schema for the file.

I think the URL you mentioned is the one which lists all the schema, one has to find the schema URL from there..

**Reply by abzcoding** [reply]
ah ok yeah that probably works

**Comment by abzcoding** [comment, marked as answer]
as mentioned by @ssbanerje this is the temporary workaround
> Ok... in the meantime I think I have found a workaround. The yamlls allows specifying a modeline:
> putting the comment at the top of a file with a URL like https://json.schemastore.org/github-workflow.json, will make the LS use that schema for the file.

```yaml
# yaml-language-server: $schema=<urlToTheSchema>
```

**Comment by kdamaski** [comment]
Hello this thread is what i was looking for. The workaround by @ssbanerje is interesting but when i load the schema like:
```
# yaml-language-server: $schema=https://json.schemastore.org/kustomization.json
```
 it has issues like not allowing "property spec". When i touch the runtime config though it works and i guess it must be handled in the lsp api.
What i would like to see though since i found a way to work with it is to be able to put the configuration in the config.lua like so:
```lua
local yaml_kube_settings = require 'lvim.lsp.providers.yamlls'.settings.yaml
yaml_kube_settings.schemaStore = {
	enable = true,
	url = "https://json.schemastore.org/kustomization.json",
}
yaml_kube_settings.schemas.kubernetes = { "*.yaml" }
```
For example the above does nothing

**Reply by kylo252** [reply]
have you tried configuring this using `:LspSettings yamlls` instead?
